### PR TITLE
Fix the AVX2 backend build

### PR DESCRIPTION
**Pull request by dkostic**
**Summary**:
For reasons that elude me, this fixed issue #574 on my machine.

**Steps**:
If your pull request consists of multiple sequential changes, please describe them here:

**Performed local tests**:
 - [ ] `lint` passing
 - [x] `tests all` passing
 - [ ] `tests bench` passing
 - [ ] `tests cbmc` passing

**Do you expect this change to impact performance**: Yes/No

If yes, please provide local benchmarking results.

### CODE REVIEW

File: mk/config.mk
Comment (hanno-becker) [inline mk/config.mk:32]: Unfortunately we can't do that without reverting further changes: We want user specified CFLAGS do overwrite default CFLAGS, but with this version it's the other way around

**Comment by hanno-becker** [comment]
The issue is lazy vs. eager evaluation of ARCH_FLAGS within CFLAGS: with := the variable is eagerly evaluated at time of definition, while with += it is lazily evaluated. This matters since auto.mk sets default ARCH_FLAGS only later.

I see two options:
1. Just move the inclusion of auto.mk upwards before the setting of CFLAGS. @dkostic can you try that please?
2. Make a copy of the original CFLAGS as (say) `ORIG_CFLAGS := $CFLAGS` and switch back to lazy evaluation for CFLAGS, but using ORIG_CFLAGS in the end.

If it works, 1. would be simpler. It depends on whether there are other places where ARCH_FLAGS is adjusted.

**Review by hanno-becker — CHANGES_REQUESTED** [review]
Thanks a lot @dkostic for spotting this. `make` apparently eludes me, too.

See the comment for what happened. For the fix we can't use the proposed change as we want to give the original CFLAGS precedence. Can you try including auto.mk earlier? Could you please also adjust the CI tests to pass `FORCE_AARCH64` and `FORCE_X86_64` accordingly, so this would have been caught?

Before we can consider merging, we need to move this PR to a non-remote branch -- this is because of some peculiarities about our EC2 CI still... (sorry!) @mkannwischer can you either give @dkostic developer rights so he can push branches, or otherwise run CI yourself when things look final?

**Comment by mkannwischer** [comment]
> Before we can consider merging, we need to move this PR to a non-remote branch -- this is because of some peculiarities about our EC2 CI still... (sorry!) @mkannwischer can you either give @dkostic developer rights so he can push branches, or otherwise run CI yourself when things look final?

Sorting out access during the holiday's is likely going to be impossible as it need approval in the TSC repository. Happy to merge this into a local branch first to run the full CI.

**Comment by mkannwischer** [comment]
> The issue is lazy vs. eager evaluation of ARCH_FLAGS within CFLAGS: with := the variable is eagerly evaluated at time of definition, while with += it is lazily evaluated. This matters since auto.mk sets default ARCH_FLAGS only later based on your system.
> 
> Some options:
> 
> 1. Just move the inclusion of auto.mk upwards before the setting of CFLAGS.
> 2. Modify CFLAGS, not ARCH_FLAGS, in auto.mk. We still need to keep ARCH_FLAGS in the CFLAGS defaults though because some test scripts set it, too.
> 3. Make a copy of the original CFLAGS as (say) `ORIG_CFLAGS := $CFLAGS` and switch back to lazy evaluation for CFLAGS, but using ORIG_CFLAGS in the end.
> 
> If it works, 2. would be simplest. Can you try @dkostic ?

I implemented (2) in #578 - let's get that merged quick and discuss any further improvment separatelty.

**Comment by hanno-becker** [comment]
Thanks @mkannwischer!

With this merged, how should this be detected in the future? Either of (a) not setting arch flags in CI, but relying on defaults, or (b) passing FORCE_xxx would have caught this -- we should probably do both. Note though that even with (a), what we benchmark is not the default build, because the latter disables LTO.

**Comment by mkannwischer** [comment]
Superseded by #578

**Comment by mkannwischer** [comment]
> Thanks @mkannwischer!
> 
> 
> 
> With this merged, how should this be detected in the future? Either of (a) not setting arch flags in CI, but relying on defaults, or (b) passing FORCE_xxx would have caught this -- we should probably do both. Note though that even with (a), what we benchmark is not the default build, because the latter disables LTO.

Let's do both.

**Pull request closed without merging**
